llvm/lib/Target/DirectX/DXILResourceAccess.cpp

```diff
@@ -28,69 +28,88 @@
-  assert(!PrevOffset && "Non-constant GEP chains not handled yet");
-
-  const DataLayout &DL = GEP->getDataLayout();
-
-  uint64_t ScalarSize = 1;
-  if (RTI.isTyped()) {
-    Type *ContainedType = RTI.getHandleTy()->getTypeParameter(0);
-    // We need the size of an element in bytes so that we can calculate the
-    // offset in elements given a total offset in bytes.
-    Type *ScalarType = ContainedType->getScalarType();
-    ScalarSize = DL.getTypeSizeInBits(ScalarType) / 8;
-  }
-
-  APInt ConstantOffset(DL.getIndexTypeSizeInBits(GEP->getType()), 0);
-  if (GEP->accumulateConstantOffset(DL, ConstantOffset)) {
-    APInt Scaled = ConstantOffset.udiv(ScalarSize);
-    return ConstantInt::get(DL.getIndexType(GEP->getType()), Scaled);
-  }
+static Value *traverseGEPOffsets(const DataLayout &DL, IRBuilder<> &Builder,
+                                 Value *Ptr, uint64_t AccessSize) {
+  Value *Offset = nullptr;
+
+  while (Ptr) {
+    if (auto *II = dyn_cast<IntrinsicInst>(Ptr)) {
+      assert(II->getIntrinsicID() == Intrinsic::dx_resource_getpointer &&
+             "Resource access through unexpected intrinsic");
+      return Offset ? Offset : ConstantInt::get(Builder.getInt32Ty(), 0);
+    }
 
-  unsigned NumIndices = GEP->getNumIndices();
-
-  // If we have a single index we're indexing into a top level array. This
-  // generally only happens with cbuffers.
-  if (NumIndices == 1)
-    return *GEP->idx_begin();
-
-  // If we have two indices, this should be a simple access through a pointer.
-  if (NumIndices == 2) {
-    auto IndexIt = GEP->idx_begin();
-    assert(cast<ConstantInt>(IndexIt)->getZExtValue() == 0 &&
-           "GEP is not indexing through pointer");
-    ++IndexIt;
-    Value *Offset = *IndexIt;
-    assert(++IndexIt == GEP->idx_end() && "Too many indices in GEP");
-    return Offset;
+    auto *GEP = dyn_cast<GetElementPtrInst>(Ptr);
+    assert(GEP && "Resource access through unexpected instruction");
+
+    unsigned NumIndices = GEP->getNumIndices();
+    uint64_t IndexScale = DL.getTypeAllocSize(GEP->getSourceElementType());
+    APInt ConstantOffset(DL.getIndexTypeSizeInBits(GEP->getType()), 0);
+    Value *GEPOffset;
+    if (GEP->accumulateConstantOffset(DL, ConstantOffset)) {
+      // We have a constant offset (in bytes).
+      GEPOffset =
+          ConstantInt::get(DL.getIndexType(GEP->getType()), ConstantOffset);
+      IndexScale = 1;
+    } else if (NumIndices == 1) {
+      // If we have a single index we're indexing into a top level array. This
+      // generally only happens with cbuffers.
+      GEPOffset = *GEP->idx_begin();
+    } else if (NumIndices == 2) {
+      // If we have two indices, this should be an access through a pointer.
+      auto IndexIt = GEP->idx_begin();
+      assert(cast<ConstantInt>(IndexIt)->getZExtValue() == 0 &&
+             "GEP is not indexing through pointer");
+      GEPOffset = *(++IndexIt);
+      assert(++IndexIt == GEP->idx_end() && "Too many indices in GEP");
```

Comment:
It doesn't seem like this assert can ever fail, since it is within an `if (NumIndices == 2)` and `IndexIt` is incremented exactly twice before the assert condition `IndexIt == GEP->idx_end()` is evaluated.